Add documents for API

cathay4t opened the pull request:
The document are intend to hold information for using libnmstate.

cathay4t commented:
Something in the document is not actually in nmstate code yet:
 * `InterfaceIPV4.ENABLE` and etc constants.
 * IPv4 and IPv6 address is not array of string yet.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+## Apply network state
+
+Nmstate support partial editing, you only need to specified the data you want
```
support > supports
.. , only the specified data is changed, the unspecified data is preserved/untouched.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+# Network State Main layout
+
+The network state is stored in a dictionary:
```
Perhaps we should explain shortly what `netinfo`, `netapplier` and `schema` represent/contain in advance.
I see the schema is explained later at line 95, perhaps consider moving it here.

Another small point about the state being `stored`. I think `represented` may be a better description of it. The state is currently stored in NM state.

EdDev commented on devel/api.md:
````diff
@@ -0,0 +1,338 @@
+```
+
+# Interface -- Basic
+A interface represents a hardware of software network interface.
````
A > An
Note that there is also the physical vs virtual interfaces, not sure which one is used more. How NM calls these?

cathay4t replied:
We don't differentiate that. I just mention some virtual Ethernet interface might not have ethernet subtree.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+# Interface -- Basic
+A interface represents a hardware of software network interface.
+Each interface will have these basic properties stored in a dictionary with
```
will have > has
Consider stored > represented
... with *the* following keys:

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+## Interface.NAME
+
+The kernel name of interface.
```
... of *the* interface

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+Each interface will have these basic properties stored in a dictionary with
+below keys. The property names are stored in `libnmstate.schema` module.
+
+## Interface.NAME
```
Are you sure this needs to be a subject and not just a member of a list?

cathay4t replied:
So that could be quickly navigate through table of content. It might help people quickly got the sections they want.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+The kernel name of interface.
+
+Vale type: `string`
```
Vale > Value
IMO you can also consider dropping the `Value`, it looks obvious.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+Possible values:
+
+ * `InterfaceType.ETHERNET` (TODO)
```
https://github.com/nmstate/nmstate/pull/346

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+   IPv4 stack is disabled.
+
+When applying network state with IPv4 disabled, all the remaining IPv4
+configurations will be ignored.
```
will be > are

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+configurations will be ignored.
+
+When showing network state, if IPv4 is disabled, all the remaining properties
+will not be included in returned network state.
```
are excluded in the ...

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+When applying network state with IPv4 disabled, all the remaining IPv4
+configurations will be ignored.
+
+When showing network state, if IPv4 is disabled, all the remaining properties
```
... *the*  network ...

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+   Ignore routes retried from DHCPv4: including gateway and
+   [classless static Route][rfc-dhcpv4-classless-route]
+
+This property only show when `ipv4['dhcp']` is `True` when querying.
```
shows

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+This property only show when `ipv4['dhcp']` is `True` when querying.
+
+This property will be ignored when `ipv4['dhcp']` is `False` when applying.
```
will be > is

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+ * `True`
+
+   Apply gateway retried from DHCPv4.
```
What do you mean by `retried`?

cathay4t replied:
retrieved

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+### IPv4 property: address
+Value type: `array of string` (TODO).
+
+The format of string is: `192.168.1.1/24`.
```
Can it be just `Fromat:` ?

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+Value type: `array of string` (TODO).
+
+The format of string is: `192.168.1.1/24`.
+When `ipv4['dhcp']` set to `True`, the address defined in this property
```
... is set ...
will also be> is also applied
... to *the* interfaces.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+   IPv6 stack is disabled.
+
+When applying network state with IPv6 disabled, all the remaining IPv6
+configurations will be ignored.
```
will be > are

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+configurations will be ignored.
+
+When showing network state, if IPv6 is disabled, all the remaining properties
+will not be included in returned network state.
```
are excluded ...

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,338 @@
+
+Due to the limitation of NetworkManager, Nmstate dose not support disabling
+IPv6 on up/active interface yet. Even with `ipv6['dhcp']` and
+`ipv6['autoconf']` both been set to `False`, interface will still get the
```
You can drop the `been`.
... , the interface is still left with an assigned IPv6 link local address.

EdDev commented on devel/use_case.md:
```diff
@@ -0,0 +1,41 @@
+Python code examples for common use cases.
+
+# Set static IP address
+
```
Can we add the use case "story" here?
Something like:
`Define an ethernet interface with two static IPv4 addresses and two static IPv6 addresses`
Note: The DHCP4/6 and AUTOCONF are optional.

EdDev reviewed (commented):
Looks very detailed, thank you.
Do you think we should add the fields documentation into the schema (I guess the description of the fields) and generate per field text?
I think tools that take a json-schema and generate that kind of text are including the type and format.

One general note: It is preferred to avoid using the future tense if possible in the documentation.
(I have checked it with one of our technical writers, just to be sure I am not inventing something) :)

cathay4t commented:
> Looks very detailed, thank you.
> Do you think we should add the fields documentation into the schema (I guess the description of the fields) and generate per field text?

I am not a fan of auto-generated python doc which turn out to be morse than manual one.

> I think tools that take a json-schema and generate that kind of text are including the type and format.

PR is welcome though I don't like it.

> 
> One general note: It is preferred to avoid using the future tense if possible in the documentation.
> (I have checked it with one of our technical writers, just to be sure I am not inventing something) :)
Done.

dankenigsberg commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+      After editing the saved output of `nmstatectl show`, `nmstatectl set`
+      command could apply the updated network state.
+
+By default, above both commands will do check whether network state is
```
"above both commands" sound odd. Maybe "The two commands above"?

maiqueb replied:
+1 on Dan's comment.

dankenigsberg commented on devel/py_example.md:
````diff
@@ -0,0 +1,351 @@
+```
+
+## Bond
+
````
An explicit `TBD` would make it clearer that there is stuff here to be documented.

maiqueb replied:
+1. It would apply to plenty of places throughout the doc, but, would make things easier going forward.

dankenigsberg reviewed (commented):
very nice addition, thanks.

qinqon commented on index.md:
```diff
@@ -28,6 +28,9 @@ pip install --user --upgrade .
+[Nmstate Installation Guide](./install.md).
+
 ### Dependencies
 In order to support specific capabilities, additional packages are required:
```
Don't wee need to add minimal NetworkManager version nmstate supports ? I am thinking on future vlan_filtering, so maybe even NetworkManager version per feature.

cathay4t replied:
Add:
```
Nmstate requires NetworkManager 1.12 or later version installed and started.
```

qinqon reviewed (changes requested):
NM version

maiqueb commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+# The nmstatectl User Guide
+
+The nmstatectl is created for user and developer who want to try out nmstate
```
... is created for users who want to try out nmstate without using libnmstate.

I mean that developers **are** users.

And the coding in python part feels unnecessary.

maiqueb commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+## Query Network State
+
+To query current network state, please use below command:
```
"the following command", or "the command below".

maiqueb commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+To commit: `nmstatectl commit <checkpoing_path>`.
+
+If no action was taken after the `TIMEOUT`(default is 60 seconds), will
```
I understand what you mean here, but I would change the wording here to highlight the user **must** commit to apply the changes - e.g. the user must commit the changes within 60 seconds, or they will be automatically rolled back. 

It means the same, but, I think the focus should be on 'user must commit'. 

Just an opinion though.

maiqueb commented on devel/api.md:
````diff
@@ -0,0 +1,990 @@
+```
+
+## Manual Transaction Control
+
````
I don't follow what happened here; is this an implicit TODO ? Is the heading level wrong ? 

What is the purpose of this section ?

cathay4t replied:
Change to `Change Network State with Manual Transaction Control`

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+## Error handling
+
+All the exceptions are stored in `libnmstate.error` and are all
```
I've read this already in https://github.com/nmstate/nmstate.github.io/pull/16/files#diff-6ed4704f70c53fa03921145d4e2259eeR99

EdDev replied:
I prefer it here and not in the intro. But there is an English problem here.

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+### `InterfaceIPv4.ENABLED`
+
+Type: `bool`
```
While I find the detail, thought, and effort you put into this amazing - kudos - I think it's a bit redundant indicating the possible values for the boolean type.

I would get rid of all the 'possible values' definition for all boolean types, since they're all quite self-explanatory.

Never the less, since you go to great lengths into explaining each of those, it's OK.

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+## `Ethernet.DUPLEX`
+
+Type: `string`, `Ethernet.FULL_DUPLEX` or `Ethernet.HALF_DUPLEX`.
```
I find the format of this description conflicting with the format you've been using above,

Type: `bool`

Possible values:

 * `True` - Apply gateway retrieved from DHCPv4.

 * `False` - Ignore gateway retrieved from DHCPv4.

The document should have one format, IMHO. Please choose one, and stick to it.

maiqueb commented on devel/api.md:
````diff
@@ -0,0 +1,990 @@
+}
+```
+
+## `LinuxBridge.GROUP_FORWARD_MASK`
````
I'm assuming these are TODOs. Am I right?

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+ * Route from DHCP or IPv6-RA.
+ * Route generated by kernel automatically. Like multicast, LAN and etc.
+
+The config routes is the static routes saved in configuration files.
```
s/is/are/

EdDev replied:
@cathay4t ping

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+The config routes is the static routes saved in configuration files.
+
+The `libnmstate.apply(desire_state)` will copy current config routes with
```
`libnmstate.apply(desired_state)`

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+* Remove specific route config entry:
+
+  Just add `Route.STATE: Route.STATE_ABSENT` to the route entry to be
```
I've just this 4/5 lines above. Is this redundant, or is there something I'm not grasping ?

cathay4t replied:
It shows the way to delete single entry of route and delete all matched route entry.

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+* Remove all route config entry that matches:
+
+  If any property is not defined in config absent route, it is treat as
+  wilecard.
```
wildcard

EdDev replied:
Not fixed.

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+  If any property is not defined in config absent route, it is treat as
+  wilecard.
+  For example, below route entry will remove all routes next hop to 'eth1'.
```
For example, the following route entry will remove all routes having the next hop parameter as 'eth1'. 

**or**

...remove all routes whose next hop is 'eth1'.

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+Type: `integer`
+
+The route table ID.
+If not define, set as `Route.USE_DEFAULT_ROUTE_TABLE` to use the `default`
```
If not defined.

maiqueb reviewed (changes requested):
Amazing improvement over the current documentation.

I think this document thoroughly describes the API, and I sincerely welcome the 'examples' doc. Neat and concise examples of common functionalities.

I've just spotted some typos, and propose minor changes.

EdDev commented on cli_guide.md:
````diff
@@ -0,0 +1,79 @@
+nmstatectl show [<interface_name>]
+```
+
+By default output format is [YAML][yaml] format, the `--json` argument could
````
YAML is the default output format. Use the `--json` argument to change the output format to JSON.

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+change the output format to [JSON][json].
+
+To limit the output state to include certain interface only, please specify
+the interface name as `nmstatectl show <interface_name>`. Please be advised,
```
We can specify multiple names.
The `--help` text says: `Show only specified interfaces (comma-separated)`.
Note that you can use UNIX shell style patterns:
```
*       matches everything
?       matches any single character
[seq]   matches any character in seq
[!seq]  matches any char not in seq
```

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+## Change Network State
+
+To change network state, you may:
```
"The network state may be changed interactively or by file:"

I think you also can drop the word "way".

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+To change network state, you may:
+
+    * Interactive way: `nmstatectl edit [<interface_name>]`
```
Perhaps specify that multiple interfaces may be given here, same as the `show` command.

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+      If there is any syntax error, you will be asked to edit again.
+
+    * File-based way: `nmstatectl set <state_file_path>`
+      After editing the saved output of `nmstatectl show`, `nmstatectl set`
```
I think it is confusing to mention here the `show` command.

smyle10 replied:
The format of file could be yaml or json. I think this should be noted.
As a new for nmstate, I thought I could only do 'nmstatectl set' with YAML file before I read README.md.

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+By default, above both commands will do check whether network state is
+identical to desired state, if not, rollback to state before `edit`/`set`
+command. To skip the verification, argument `--no-verify` could be used.
```
How about: "Use the `--no-verify` argument to skip the verification." (not a must)

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+This document holds all the API detail of the `libnmstate` Python module.
+For quick start, you may refer to [Python code examples][py_example]
+
+The `libnmstate` contains:
```
The `libnmstate` package exposes the following API:

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+    * `apply()` -- [Apply network state][apply]
+    * `commit()` -- [Commit a transaction][manual]
+    * `rollback()` -- [Rollback a transaction][manual]
+ * [All exception classes][error] are stored in `libnmstate.error` and
```
Can we show the submodules in a similar way to the functions?
Submodules:
`error` -- [Exceptions][error]
`schema` -- Schema Constants

I think that the error hierarchy can be explained later on and not here at the into.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+import libnmstate
+
+net_state = libnmstate.show()
+print(json.dumps(net_state, indent=4))
```
I think that it is less interesting to print the output here.
Maybe show some basic usage of net_state is better, like:
`ifaces_names = [iface_state[Interface.NAME] for iface_state in net_state[Interface.KEY]]`

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+Nmstate supports partial editing, only the specified data is changed, the
+unspecified data is preserved/untouched.
+
```
Maybe we should add an exception clause. Like `[exceptions apply][...]`
This is incorrect for IP addresses, bond slaves, bridge ports (at the moment a limitation we need to fix).

EdDev replied:
What about this one @cathay4t ?

cathay4t replied:
Added:

Some properties(e.g. `InterfaceIPv4.ADDRESS` ) do not support partial editing,
please refer to their document for detail.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+### NmstateConflictError
+
+Something else is already editing the network state via Nmstate.
```
How about: "An existing Nmstate transaction is in progress."

Note that if I am not mistaken, if the transaction is executed with a timeout and a commit is expected to cancel the auto-rollback, the timeout duration is considered part of the transaction progress.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+### NmstateVerificationError
+
+After applied desired state, current state does not match desired state for
```
I think that the order is important here: The desired state does not match the refreshed current state.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+The YAML schema file is placed at `libnmstate/schemas/operational-state.yaml`.
+
+In stead of using hard coded string, please use classes and constants in
```
Prefer using the constants defined in `libnmstate.schema`.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+In stead of using hard coded string, please use classes and constants in
+`libnmstate.schmea`.
+
+# Limitations
```
I think this belongs to a separate file in the source repo. Changing the user guide when a bug is fixed is not that common.
But we could place a link here to the limitation text.

cathay4t replied:
https://github.com/nmstate/nmstate/pull/500

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+# Basic Interface
+
+An interface represents a hardware or software network interface.
+Each interface will has these basic properties represented as a dictionary.
```
Drop the `will`.
Consider removing the redundant "dictionary" text. This section can focus on the content.
(and it being a dict is obvious from the snippets)

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+
+# Ethernet
+
+Besides basic interface properties, each Ethernet interface state also contains
```
The `ethernet` type interface specific properties are placed under `Ethernet.CO..`.

Consider using this format for all the others.

EdDev commented on devel/api.md:
```diff
@@ -0,0 +1,990 @@
+Besides basic interface properties, each Ethernet interface state also contains
+a dictionary saved in key `Ethernet.CONFIG_SUBTREE`.
+
+For virtual interface like `virtio_net` or `veth`, there is no
```
Isn't this obvious? These are other types.

cathay4t replied:
But they are list as `type: ethernet`, so a document should explain why there is `Ethernet.CONFIG_SUBTREE` for them.

smyle10 commented on cli_guide.md:
```diff
@@ -0,0 +1,79 @@
+
+    * Interactive way: `nmstatectl edit [<interface_name>]`
+      The `nmstatectl` will invoke the text editor defined by eviroment
+      variable `$EDITOR` for editing the network state in [YAML][yaml] format.
```
in [YAML][yaml] format.
---- it shows as literal  "[YAML][yaml]" in web-view when I view the original file, not a formatted link.

cathay4t replied:
Works for me. Please check again.

EdDev replied:
I think nothing changed.

cathay4t replied:
It works well without changes.

EdDev commented:
@cathay4t , I see that several comments have not been responded.

smyle10 commented on devel/api.md:
```diff
@@ -0,0 +1,1079 @@
+ * `InterfaceState.ABSENT`
+
+   Remove interface.
+
```
No 'unknown' anymore? I see a 'unknown' value in https://github.com/nmstate/nmstate/blob/master/libnmstate/schemas/operational-state.yaml

cathay4t replied:
Fixed in https://github.com/nmstate/nmstate/pull/501

smyle10 commented on devel/api.md:
```diff
@@ -0,0 +1,1079 @@
+
+The MAC address of interface.
+
+Type: `String` in the format of `EA:60:E4:08:17:F1`.
```
Is it case sensitive?

cathay4t replied:
Case insensitive. Document added.

EdDev commented:
@cathay4t, can you have another look at the feedback so far so we can take this in?

cathay4t commented:
@EdDev New patch uploaded with feedback applied.

EdDev commented on cli_guide.md:
```diff
@@ -0,0 +1,99 @@
+The network state may be changed interactively or by file:
+
+    * Interactive: `nmstatectl edit [<interface_names>]`
+      The `nmstatectl` will invoke the text editor defined by eviroment
```
Please fix `eviroment`

cathay4t commented:
@EdDev New patch uploaded with feedback applied.

maiqueb commented on devel/api.md:
````diff
@@ -0,0 +1,1074 @@
+}
+```
+
+The running routes contains route entires in kernel, it might contains:
````
The running routes contains route **entries** in kernel, **which** might **contain**

maiqueb commented on devel/api.md:
````diff
@@ -0,0 +1,1074 @@
+```
+
+The running routes contains route entires in kernel, it might contains:
+ * Route from DHCP or IPv6-RA.
````
shouldn't 'route' be plural ? i.e. routes ?

maiqueb commented on devel/api.md:
```diff
@@ -0,0 +1,1074 @@
+
+The running routes contains route entires in kernel, it might contains:
+ * Route from DHCP or IPv6-RA.
+ * Route generated by kernel automatically. Like multicast, LAN and etc.
```
ditto

cathay4t committed:
Add documents for API, nmstatectl and installation

The API document is intend to hold information for using libnmstate.
The nmstate user guide is intend to hold information for using
nmstatectl.

Signed-off-by: Gris Ge <fge@redhat.com>
